repositories/strategies/spacedOrderStrategy.js

```diff
@@ -46,13 +58,48 @@ const getOrUpdateOrderForInsert = async (trx, sectionId, id, position) => {
   let right = getOr(maxOrder, "order", pages[position]);
 
   if (valuesHaveConverged(left, right)) {
-    await makeSpaceForInsert(trx, sectionId, left);
+    await makeSpaceForPageInsert(trx, sectionId, left);
+    right += SPACING;
+  }
+
+  return calculateMidPoint(left, right);
+};
+
+const getOrUpdateOrderForSectionInsert = async (
+  trx,
+  questionnaireId,
+  id,
+  position
+) => {
+  const sections = reject(
+    { id },
+    await getSectionsByQuestionnaire(trx, questionnaireId)
+  );
+  const maxOrder = getMaxOrder(sections) + SPACING;
+
+  position = isNil(position)
+    ? sections.length
+    : clamp(0, sections.length, position);
+
+  if (isEmpty(sections)) {
+    return SPACING;
+  }
+  if (position === sections.length) {
+    return maxOrder;
+  }
+
+  let left = getOr(0, "order", sections[position - 1]);
+  let right = getOr(maxOrder, "order", sections[position]);
+
+  if (valuesHaveConverged(left, right)) {
+    await makeSpaceForSectionInsert(trx, questionnaireId, left);
```

Comment:
There's some duplication of logic here that could be eliminated.

```js
const makeSpaceForInsert = (trx, tableName, parentId, order) => {
  const key = tableName === "Sections" ? "QuestionnaireId" : "SectionId";
  return trx(tableName)
    .where({ [key]: parentId })
    .andWhere("order", ">", order)
    .increment("order", SPACING);
};

const getOrUpdateOrderForInsert = async (
  trx,
  collection,
  type,
  parentId,
  position
) => {
  const maxOrder = getMaxOrder(collection) + SPACING;

  position = isNil(position)
    ? collection.length
    : clamp(0, collection.length, position);

  if (isEmpty(collection)) {
    return SPACING;
  }
  if (position === collection.length) {
    return maxOrder;
  }

  let left = getOr(0, "order", collection[position - 1]);
  let right = getOr(maxOrder, "order", collection[position]);

  if (valuesHaveConverged(left, right)) {
    await makeSpaceForInsert(trx, type, parentId, left);
    right += SPACING;
  }

  return calculateMidPoint(left, right);
};

const getOrUpdateOrderForPageInsert = async (trx, sectionId, id, position) => {
  const pages = reject({ id }, await getPagesBySection(trx, sectionId));

  return getOrUpdateOrderForInsert(trx, pages, "Pages", sectionId, position);
};

const getOrUpdateOrderForSectionInsert = async (
  trx,
  questionnaireId,
  id,
  position
) => {
  const sections = reject(
    { id },
    await getSectionsByQuestionnaire(trx, questionnaireId)
  );

  return getOrUpdateOrderForInsert(
    trx,
    sections,
    "Sections",
    questionnaireId,
    position
  );
};
```

Reply:
You could possibly get rid of `getOrUpdateOrderForPageInsert` and `getOrUpdateOrderForSectionInsert` altogether.